### PR TITLE
[type] Support bit-level load and store

### PR DESCRIPTION
**Pull request by Hanke98**
Related issue = #1905 #1968 
This pr aims to implement bit-level operations including bit-level GlobalLoad, GLobalStore, GlobalPtr, SNodeLookUp, GetChild. Currently I make the `SNodeLookUp` and `GetChild` work and the rest are works in progress.

 @yuanming-hu, could you please have a look on these changes when you are available?  Your advices will be highly appreciated ! Thanks!

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/struct/struct_llvm.cpp [hunk 1]
```diff
@@ -25,8 +25,8 @@ StructCompilerLLVM::StructCompilerLLVM(Program *prog, Arch arch)
 void StructCompilerLLVM::generate_types(SNode &snode) {
   TI_AUTO_PROF;
   auto type = snode.type;
-  if (snode.is_bit_level)
-    return;
+//  if (snode.is_bit_level)
+//    return;
```
Comment (Hanke98) [inline taichi/struct/struct_llvm.cpp:29]: According to my understanding, we might also need to execute this function for bit-level SNodes.

File: taichi/struct/struct_llvm.cpp [hunk 2]
```diff
@@ -37,12 +37,17 @@ void StructCompilerLLVM::generate_types(SNode &snode) {
...
+    } else {
+      DataType container_primitive_type(snode.physical_type);
+      ch_types.push_back(tlctx->get_data_type(container_primitive_type));
     }
   }
 
+  // if we don't push any thing into `ch_types` for bit_level SNode
+  // `ch_type` will be  a StructType with no element in it.
   auto ch_type =
       llvm::StructType::create(*ctx, ch_types, snode.node_type_name + "_ch");
```
Comment (Hanke98) [inline taichi/struct/struct_llvm.cpp:52]: As I comment here, bit-level SNodes need to know their children types.

File: taichi/struct/struct_llvm.cpp (hunk 1 above)
Comment (yuanming-hu) [reply taichi/struct/struct_llvm.cpp:29]: Actually, I don't think we should comment out these lines. Note that for `bit_level` SNodes we never actually generate their LLVM types (since LLVM doesn't really have corresponding types).

File: taichi/struct/struct_llvm.cpp (hunk 2 above)
Comment (yuanming-hu) [reply taichi/struct/struct_llvm.cpp:52]: In fact, they don't need to know their children types - and LLVM doesn't have corresponding mechanisms to generate these types. That's why we need bit pointers.

**Review by yuanming-hu — COMMENTED** [review]
Thanks for sharing with me the progress! In fact, what we really need to do here is to modify the LLVMCodegen class to generate code for bit pointers. I believe not much should be done on other files.

To do that, as we discussed before, create an LLVM `alloca` of type `struct {i8 *, i32}` variable (to store the bit pointers) for the statements that involve global pointers (e.g., `GetCh`) in the LLVM code generator. Then for `GlobalLoad/StoreStmt`, if the input pointer is a bit pointer (you will know that at compile time from the input pointer return type `is_bit_pointer()` field), generate an LLVM `load` instruction with type `pointer->ret_type->physical_type` and then use bit SHR/AND operations to extract the corresponding bits.

File: taichi/struct/struct_llvm.cpp (hunk 1 above)
Comment (Hanke98) [reply taichi/struct/struct_llvm.cpp:29]: Yes, I didn't comment them out at beginning, but the program will crash when visiting `SNodeLookUpStmt`, and I thought it might because we don't generate the types for bit-level SNodes.

**Comment by Hanke98** [comment]
> Thanks for sharing with me the progress! In fact, what we really need to do here is to modify the LLVMCodegen class to generate code for bit pointers. I believe not much should be done on other files.
> 
> To do that, as we discussed before, create an LLVM `alloca` of type `struct {i8 *, i32}` variable (to store the bit pointers) for the statements that involve global pointers (e.g., `GetCh`) in the LLVM code generator. Then for `GlobalLoad/StoreStmt`, if the input pointer is a bit pointer (you will know that at compile time from the input pointer return type `is_bit_pointer()` field), generate an LLVM `load` instruction with type `pointer->ret_type->physical_type` and then use bit SHR/AND operations to extract the corresponding bits.

Thanks for your very detailed suggestions. I still feel a little confused about SNodes, and let's discuss them in the next meeting. For now, let me focus on `LLVMCodegen` class : )

File: taichi/codegen/codegen_llvm.cpp [hunk 3]
```diff
@@ -1187,7 +1213,8 @@ void CodeGenLLVM::visit(SNodeLookupStmt *stmt) {
   } else if (snode->type == SNodeType::dense ||
              snode->type == SNodeType::pointer ||
              snode->type == SNodeType::dynamic ||
-             snode->type == SNodeType::bitmasked) {
+             snode->type == SNodeType::bitmasked ||
+             snode->type == SNodeType::bit_struct) {
```
Comment (Hanke98) [inline taichi/codegen/codegen_llvm.cpp:1217]: I add this line because when I run script `test_bit_struct.py`, the visiting of `SNodeLookUpStmt` will crash. I might not use a right method to handle this problem, but I think we can ignore this for now, I will take care of it after the `GetChStmt`,`GlobalLoadStmt` and `GlobalStoreStmt` are right.

**Comment by Hanke98** [comment]
> To do that, as we discussed before, create an LLVM `alloca` of type `struct {i8 *, i32}` variable (to store the bit pointers) for the statements that involve global pointers (e.g., `GetCh`) in the LLVM code generator. Then for `GlobalLoad/StoreStmt`, if the input pointer is a bit pointer (you will know that at compile time from the input pointer return type `is_bit_pointer()` field), generate an LLVM `load` instruction with type `pointer->ret_type->physical_type` and then use bit SHR/AND operations to extract the corresponding bits.

Hi, @yuanming-hu, I modify the visiting of `GetChStmt` and `GlobalLoadStmt` in code generator. Please help me to know if I am on the right way? Thank you very much!

File: taichi/codegen/codegen_llvm.cpp [hunk 4]
```diff
@@ -1201,14 +1228,44 @@ void CodeGenLLVM::visit(SNodeLookupStmt *stmt) {
...
+    auto byte_ptr = builder->CreateBitCast(llvm_val[stmt->input_ptr],
+                           llvm::PointerType::getInt8PtrTy(*llvm_context));
+
+    builder->CreateStore(byte_ptr, builder->CreateGEP(bit_ptr_struct,
+                                                      {tlctx->get_constant(0), tlctx->get_constant(0)}));
+    // 4. store `offset`
+    // temporarily use 0 for offset.
+    // TODO: correct offset.
+    builder->CreateStore(tlctx->get_constant(0), builder->CreateGEP(bit_ptr_struct,
+                                                                    {tlctx->get_constant(0), tlctx->get_constant(1)}));
```
Comment (Hanke98) [inline taichi/codegen/codegen_llvm.cpp:1255]: I am not sure how do we get the right offset here. So I use zero currently.

File: taichi/codegen/codegen_llvm.cpp (hunk 3 above)
Comment (yuanming-hu) [reply taichi/codegen/codegen_llvm.cpp:1217]: Actually for `bit_struct`, you can simply return `parent` as the `llvm_val`.

File: taichi/codegen/codegen_llvm.cpp [hunk 5]
```diff
@@ -1075,8 +1079,26 @@ void CodeGenLLVM::visit(GlobalStoreStmt *stmt) {
 void CodeGenLLVM::visit(GlobalLoadStmt *stmt) {
   int width = stmt->width();
   TI_ASSERT(width == 1);
-  llvm_val[stmt] = builder->CreateLoad(tlctx->get_data_type(stmt->ret_type),
-                                       llvm_val[stmt->ptr]);
+  if (auto cit = stmt->ret_type->cast<CustomIntType>()) {
```
Comment (yuanming-hu) [inline taichi/codegen/codegen_llvm.cpp:1082]: Here I guess testing `stmt->ptr->ret_type->cast<PointerType>()->is_bit_pointer()` is better here?

File: taichi/codegen/codegen_llvm.cpp [hunk 6]
```diff
@@ -1201,14 +1228,44 @@ void CodeGenLLVM::visit(SNodeLookupStmt *stmt) {
...
+    // struct bit_pointer {
+    //    i8* byte_ptr;
+    //    i32 offset;
+    // };
+    auto struct_type = llvm::StructType::get(*llvm_context, {llvm::Type::getInt8PtrTy(*llvm_context),
+                                                            llvm::Type::getInt32Ty(*llvm_context)});
+    // 2. alloca the bit pointer struct
+    auto bit_ptr_struct = create_entry_block_alloca(struct_type);
+
+    // 3. store the `input_ptr` in to `bit_struct`
```
Comment (yuanming-hu) [inline taichi/codegen/codegen_llvm.cpp:1245]:
```suggestion
    // 3. store `input_ptr` into `bit_ptr_struct`
```

File: taichi/codegen/codegen_llvm.cpp (hunk 4 above)
Comment (yuanming-hu) [reply taichi/codegen/codegen_llvm.cpp:1255]: `BitStructType::get_member_bit_offset` may be the right thing to look at. See also `GetChStmt::chid` :-)

**Review by yuanming-hu — COMMENTED** [review]
Great! I think we are mostly on the right direction. `BitStructMeta` seems not necessary, but the `GetCh` and `GlobalLoad` codegen looks good!

File: taichi/codegen/codegen_llvm.cpp (hunk 3 above)
Comment (yuanming-hu) [reply taichi/codegen/codegen_llvm.cpp:1217]: (We may need a `BitArrayMeta` later for bitmasked bit arrays, but for now and for `BitStruct`, we can simply stick to the simple solution above)

File: taichi/codegen/codegen_llvm.cpp (hunk 5 above)
Comment (Hanke98) [reply taichi/codegen/codegen_llvm.cpp:1082]: Yes, you’re right 👍

**Comment by Hanke98** [comment]
> Great! I think we are mostly on the right direction. `BitStructMeta` seems not necessary, but the `GetCh` and `GlobalLoad` codegen looks good!

Good, thanks a lot for your careful review and detailed suggestions. So I will go on implementing the rest of bit-level statements after applying your advices.

File: taichi/codegen/codegen_llvm.cpp [hunk 7]
```diff
@@ -1069,14 +1073,56 @@ void CodeGenLLVM::visit(GlobalStoreStmt *stmt) {
...
+    // 1. load byte pointer
+    auto byte_ptr_in_bit_struct = builder->CreateGEP(llvm_val[stmt->ptr], {tlctx->get_constant(0), tlctx->get_constant(0)});
+    auto byte_ptr = builder->CreateLoad(byte_ptr_in_bit_struct);
+    assert(byte_ptr->getType()->getPointerElementType()->isIntegerTy(8));
+    auto bit_level_container = builder->CreateLoad(builder->CreateBitCast(byte_ptr, llvm::Type::getInt32PtrTy(*llvm_context)));
+
+    // 2. load bit offset
+    auto bit_offset_in_bit_struct = builder->CreateGEP(llvm_val[stmt->ptr], {tlctx->get_constant(0), tlctx->get_constant(1)});
+    auto bit_offset = builder->CreateLoad(bit_offset_in_bit_struct);
+    assert(bit_offset->getType()->isIntegerTy(32));
```
Comment (Hanke98) [inline taichi/codegen/codegen_llvm.cpp:1111]: I will move these repeating lines(same in GlobalStoreStmt and GlobalLoad) into a new function.

File: taichi/codegen/codegen_llvm.cpp (hunk 5 above)
Comment (Hanke98) [reply taichi/codegen/codegen_llvm.cpp:1100]: > Here I guess testing `stmt->ptr->ret_type->cast<PointerType>()->is_bit_pointer()` is better here?

I tried to test `stmt->ptr->ret_type->cast<PointerType>()->is_bit_pointer()`, but it seems always False. Maybe we need to have a look at when do we set ret_type bit-level.

**Comment by Hanke98** [comment]
Hi, @yuanming-hu. A very basic version of GlobalLoad and GlobalStore is implemented. Could you please have another review when you are free? Thanks again! :-)

File: misc/test_bit_struct.py [hunk 8]
```diff
@@ -42,9 +42,14 @@ def test_simple_singleton():
 
     @ti.kernel
     def foo():
-        print(x[None])
+        x[None] = 2**13-1
+        print('x: (2**13-1)', x[None])
```
Comment (yuanming-hu) [inline misc/test_bit_struct.py:46]: You can `ti.init(debug=True)` and then use `assert` in Taichi kernels. Then you don't need to eyeballing the outputs anymore.

File: taichi/codegen/codegen_llvm.cpp [hunk 9]
```diff
@@ -209,7 +209,11 @@ std::unique_ptr<RuntimeObject> CodeGenLLVM::emit_struct_meta_object(
     meta =
         std::make_unique<RuntimeObject>("BitmaskedMeta", this, builder.get());
     emit_struct_meta_base("Bitmasked", meta->ptr, snode);
-  } else {
+  } else if (snode->type == SNodeType::bit_struct) {
+    meta = std::make_unique<RuntimeObject>("BitStructMeta", this, builder.get());
+    emit_struct_meta_base("BitStruct", meta->ptr, snode);
```
Comment (yuanming-hu) [inline taichi/codegen/codegen_llvm.cpp:214]: Do we still need this?

File: taichi/codegen/codegen_llvm.cpp [hunk 10]
```diff
@@ -1069,14 +1073,56 @@ void CodeGenLLVM::visit(GlobalStoreStmt *stmt) {
...
   TI_ASSERT(width == 1);
-  llvm_val[stmt] = builder->CreateLoad(tlctx->get_data_type(stmt->ret_type),
-                                       llvm_val[stmt->ptr]);
+  TI_INFO("stmt->ptr->ret_type->cast<PointerType>()->is_bit_pointer(): {}", stmt->ptr->ret_type->cast<PointerType>()->is_bit_pointer());
+  if (auto cit = stmt->ret_type->cast<CustomIntType>()) {
+    TI_INFO("bit-level global load {}", cit->to_string());
+    // 1. load byte pointer
+    auto byte_ptr_in_bit_struct = builder->CreateGEP(llvm_val[stmt->ptr], {tlctx->get_constant(0), tlctx->get_constant(0)});
+    auto byte_ptr = builder->CreateLoad(byte_ptr_in_bit_struct);
+    assert(byte_ptr->getType()->getPointerElementType()->isIntegerTy(8));
```
Comment (yuanming-hu) [inline taichi/codegen/codegen_llvm.cpp:1105]: `TI_ASSERT`

**Review by yuanming-hu — COMMENTED** [review]
Very nice! I took a quick look and I believe the implementation is mostly good! I think now we can clean up the code a little bit, and add 1-2 more test cases. I'll take another look after the code is cleaned up :-)

In a future PR, we can implement CustomInt SNode read/write in Python-scope.

File: taichi/transforms/compile_to_offloads.cpp [hunk 11]
```diff
@@ -102,9 +102,12 @@ void compile_to_offloads(IRNode *ir,
...
 
-  irpass::cfg_optimization(ir, false);
-  print("Optimized by CFG");
-  irpass::analysis::verify(ir);
+  // comment out `cfg_optimization` temporarily
+  // to avoid optimizing some important operations
+  // when testing bit-level GlobalLoad/Store
+//  irpass::cfg_optimization(ir, false);
+//  print("Optimized by CFG");
+//  irpass::analysis::verify(ir);
```
Comment (TH3CHARLie) [inline taichi/transforms/compile_to_offloads.cpp:110]: I believe this should be reset when the PR is ready for its final-round review.

Comment (Hanke98) [reply taichi/transforms/compile_to_offloads.cpp:110]: Yes, you are right, I did this to avoid some storing or loading operations being optimized when testing. BTW, do you know how to skip some IR optimization pass other than simply commenting them out?

**Comment by Hanke98** [comment]
Finally, I managed to support the bit-level signed/unsigned custom integer load and store. One can now use `ti test bit_struct`  to test the changes.

File: misc/test_bit_struct.py [hunk 12]
```diff
@@ -27,22 +27,28 @@ def foo():
...
 
     @ti.kernel
     def foo():
-        print(x[None])
+        x[None] = 2**12-1
+        assert(x[None] == 2**12-1)
+        y[None] = 2**14-1
+        assert(y[None] == 2**14-1)
+        z[None] = -(2**3)
+        assert(z[None] == -2**3)
```
[the remark this answers is not in the document]
Comment (yuanming-hu) [reply misc/test_bit_struct.py:51]: Let's also make sure `test_simple_array` passes in the tests. This file can be then removed.

File: taichi/transforms/type_check.cpp [hunk 13]
```diff
@@ -264,9 +267,20 @@ class TypeCheck : public IRVisitor {
...
+            return insert_type_cast_before(stmt, hs, get_data_type<uint32>());
+        }
+        return hs;
+      };
+      stmt->lhs = promote_custom_int_type(stmt, stmt->lhs);
+      stmt->rhs = promote_custom_int_type(stmt, stmt->rhs);
       auto ret_type = promoted_type(stmt->lhs->ret_type, stmt->rhs->ret_type);
       if (ret_type != stmt->lhs->ret_type) {
-        // promote rhs
+        // promote lhs
```
Comment (yuanming-hu) [inline taichi/transforms/type_check.cpp:283]: Thanks! Nice catch.

File: taichi/runtime/llvm/runtime.cpp [hunk 14]
```diff
@@ -1550,6 +1550,16 @@ void stack_push(Ptr stack, size_t max_num_elements, std::size_t element_size) {
...
 #include "internal_functions.h"
+
+void set_partial_bits_b32(u32*ptr, u32 offset, u32 bits, u32 value) {
+  u32 mask = ~(u32)0;
+  mask = ((mask << offset) >> (32 - offset - bits));
+  u32 new_value = 0;
+  do {
+    u32 old_value = *ptr;
+    new_value = (old_value & (~mask)) | (value << offset);
+  } while (!atomic_exchange_u32(ptr, new_value));
```
Comment (yuanming-hu) [inline taichi/runtime/llvm/runtime.cpp:1561]: I believe here we should use `__atomic_compare_exchange` instead of `atomic_exchange_u32`. See

https://github.com/taichi-dev/taichi/blob/dea88d0531e0097a355b024753c2850ca7622d56/taichi/runtime/llvm/atomic.h#L63-L86

File: taichi/runtime/llvm/runtime.cpp [hunk 15]
```diff
@@ -1550,6 +1550,16 @@ void stack_push(Ptr stack, size_t max_num_elements, std::size_t element_size) {
 }
 
 #include "internal_functions.h"
+
+void set_partial_bits_b32(u32*ptr, u32 offset, u32 bits, u32 value) {
+  u32 mask = ~(u32)0;
+  mask = ((mask << offset) >> (32 - offset - bits));
```
Comment (yuanming-hu) [inline taichi/runtime/llvm/runtime.cpp:1556]: I'm a little confused here - if we are counting 0 from LSB, the correct mask should have `1`s starting from `offset` to `offset + bits`. That doesn't seem to be the case here.

One quick way to achieve the mask above seems to be `(((u32)1 << bits) - 1) << offset)`.

File: taichi/codegen/codegen_llvm.cpp (hunk 5 above)
Comment (yuanming-hu) [reply taichi/codegen/codegen_llvm.cpp:1095]: OK, sounds good. I'll leave this task to you :-)

File: taichi/codegen/codegen_llvm.cpp [hunk 16]
```diff
@@ -1069,14 +1077,46 @@ void CodeGenLLVM::visit(GlobalStoreStmt *stmt) {
...
+    auto bit_level_container = builder->CreateLoad(builder->CreateBitCast(byte_ptr, llvm::Type::getInt32PtrTy(*llvm_context)));
+    // 2. bit shifting
+    //    first left shift `32 - (offset + num_bits)`
+    //    then right shift `32 - num_bits`
+    auto bit_end = builder->CreateAdd(bit_offset,
+                                      tlctx->get_constant(cit->get_num_bits()));
+    auto left = builder->CreateAdd(tlctx->get_constant(32),
+                                   builder->CreateNeg(bit_end));
+    auto right = builder->CreateAdd(tlctx->get_constant(32),
+                                    tlctx->get_constant(-cit->get_num_bits()));
```
Comment (yuanming-hu) [inline taichi/codegen/codegen_llvm.cpp:1108]: Use `builder->CreateSub`?

**Review by yuanming-hu — COMMENTED** [review]
Thanks! The `GetCh` part is well done! I'll take another look after the issues are addressed. Please re-request reviews from @TH3CHARLie and me after you fix the issues.

File: taichi/transforms/compile_to_offloads.cpp [hunk 17]
```diff
@@ -102,9 +102,12 @@ void compile_to_offloads(IRNode *ir,
...
 
-  irpass::cfg_optimization(ir, false);
-  print("Optimized by CFG");
-  irpass::analysis::verify(ir);
+  // comment out `cfg_optimization` temporarily
+  // to avoid optimizing some important operations
+  // when testing bit-level GlobalLoad/Store
+//    irpass::cfg_optimization(ir, false);
+//    print("Optimized by CFG");
+//    irpass::analysis::verify(ir);
```
[the remark this answers is not in the document]
Comment (Hanke98) [reply taichi/transforms/compile_to_offloads.cpp]: Thanks, this is what I need!

File: misc/test_bit_struct.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Hanke98) [reply misc/test_bit_struct.py]: > It would be nice to move the `asserts` into a separate `bar` function. If the read and writes happen in the same kernel the optimizer may simply do store-forwarding and then the codegen is not really tested.

Thanks for telling me. It's a good idea to separate the the function. That avoids some problems caused by optimizers. Let me write the test functions in the new test file in this way.

**Comment by Hanke98** [comment]
> Thanks! The `GetCh` part is well done! I'll take another look after the issues are addressed. Please re-request reviews from @TH3CHARLie and me after you fix the issues.

A lot of thanks for your careful review and advices and sorry for some unnecessary mistakes of mine. This time I have fixed the issue you mentioned above and please have another review~

File: tests/python/test_bit_struct.py [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    cu14 = ti.type_factory_.get_custom_int_type(14, False)
+    ci5 = ti.type_factory_.get_custom_int_type(5, True)
+
+    x = ti.field(dtype=ci13)
+    y = ti.field(dtype=cu14)
+    z = ti.field(dtype=ci5)
+
+    test_case_np = np.array([[2**12 - 1, 2**14 - 1, -(2**3)],
+                             [2**11 - 1, 2**13 - 1, -(2**2)], [0, 0, 0]],
+                            dtype=np.int32)
```
Comment (yuanming-hu) [inline tests/python/test_bit_struct.py:47]: It's nice to test the boundary values of those CustomIntTypes. How about adding a handy extension in this test case to cover another group of `test_case_np` with non-boundary values?

**Review by yuanming-hu — APPROVED** [review]
Awesome! LGTM now. If that makes sense, I would suggest adding another group of `test_case_np`. Everything else looks great!

**Comment by TH3CHARLie** [comment]
Congratulations and great job! I think I do not have enough context to do a full review for this PR as I am actually reviewing #1968 myself to fill my gaps. This PR is also an important one for me to learn and move foward

File: tests/python/test_bit_struct.py (hunk 18 above)
Comment (Hanke98) [reply tests/python/test_bit_struct.py:47]: Sure! I will add some other test cases.
